qiskit/quantum_info/operators/symplectic/sparse_pauli_op.py

```diff
@@ -0,0 +1,564 @@
+from qiskit.quantum_info.operators.operator import Operator
+from qiskit.quantum_info.operators.symplectic.pauli_table import PauliTable
+from qiskit.quantum_info.operators.symplectic.pauli_utils import pauli_basis
+from qiskit.quantum_info.operators.custom_iterator import CustomIterator
+
+
+class SparsePauliOp(BaseOperator):
+    """Sparse N-qubit operator in a Pauli basis representation.
+
+    This is a sparse representation of an N-qubit matrix
+    :class:`~qiskit.quantum_info.Operator` in terms of N-qubit
+    :class:`~qiskit.quantum_info.PauliTable` and complex coefficients.
+
+    It can be used for performing operator arithmetic for hundred of qubits
+    if the number of non-zero Pauli basis terms is sufficiently small.
+
+    The Pauli basis components are stored as a
+    :class:`~qiskit.quantum_info.PauliTable` object and can be accessed
+    using the :attr:`~SparsePauliOp.pauli` attribute. The coefficients
+    are stored as a complex Numpy array vector and can be accessed using
+    the :attr:`~SparsePauliOp.coeffs` attribute.
+    """
+
+    def __init__(self, data, coeffs=None):
+        """Initialize an operator object."""
+        if isinstance(data, SparsePauliOp):
+            pauli = data._pauli
+            coeffs = data._coeffs
+        else:
+            pauli = PauliTable(data)
+            if coeffs is None:
+                coeffs = np.ones(pauli.size, dtype=np.complex)
+        # Initialize PauliTable
+        self._pauli = pauli
+
+        # Initialize Coeffs
+        self._coeffs = np.asarray(coeffs, dtype=complex)
+        if self._coeffs.shape != (self._pauli.size, ):
+            raise QiskitError("coeff vector is incorrect shape for number"
+                              " of Paulis {} != {}".format(self._coeffs.shape,
+                                                           self.pauli.size))
+        # Initialize BaseOperator
+        super().__init__(pauli._input_dims, pauli._output_dims)
+
+    def __repr__(self):
+        prefix = 'SparsePauliOp('
+        pad = len(prefix) * ' '
+        return '{}{},\n{}coeffs={})'.format(
+            prefix, np.array2string(
+                self.pauli.array, separator=', ', prefix=prefix),
+            pad, np.array2string(self.coeffs, separator=', '))
+
+    def __eq__(self, other):
+        """Check if two SparsePauliOp operators are equal"""
+        return (super().__eq__(other)
+                and np.allclose(self.coeffs, other.coeffs)
+                and self.pauli == other.pauli)
+
+    # ---------------------------------------------------------------------
+    # Data accessors
+    # ---------------------------------------------------------------------
+
+    @property
+    def size(self):
+        """The number of Pauli of Pauli terms in the operator."""
+        return self._pauli.size
+
+    def __len__(self):
+        """Return the size."""
+        return self.size
+
+    @property
+    def pauli(self):
```

Comment:
Reason this is "pauli" and not "paulis"?

Reply:
Might be better I rename this to `table`